### PR TITLE
Use reductions with reused tmp storage in solver

### PR DESCRIPTION
**Pull request by upsj**
This PR uses the new interface for Dense reduction kernels from #990 in all Krylov solvers that use them

### CODE REVIEW

File: core/solver/bicg.cpp [hunk 1]
```diff
@@ -129,6 +129,8 @@ void Bicg<ValueType>::apply_dense_impl(const matrix::Dense<ValueType>* dense_b,
 
     auto exec = this->get_executor();
 
+    Array<char> reduction_tmp{exec};
```
Comment (pratikvn) [inline core/solver/bicg.cpp:132]: I guess the idea is to move this to one of the base classes in the value semantics (solver base class) PRs ?

**Review by pratikvn — APPROVED** [review]
LGTM!

File: core/solver/bicgstab.cpp [hunk 2]
```diff
@@ -111,6 +111,8 @@ void Bicgstab<ValueType>::apply_dense_impl(
 
     auto exec = this->get_executor();
 
+    Array<char> reduction_tmp{exec};
```
Comment (yhmtsai) [inline core/solver/bicgstab.cpp:114]: move it to mutable class member? in case we need reusing these in the multigrid or something else.

Comment (upsj) [reply core/solver/bicgstab.cpp:114]: That should be handled later IMO, together with #805.

**Review by yhmtsai — APPROVED** [review]

**Comment by upsj** [comment]
rebase!

**Pull request merged**